apps/builder/app/builder/features/settings-panel/variables.tsx

```diff
@@ -0,0 +1,362 @@
+          overflow: "hidden",
+          p: theme.spacing[9],
+          gap: theme.spacing[5],
+        }}
+      >
+        <Flex direction="column" css={{ gap: theme.spacing[3] }}>
+          <Label htmlFor={nameId}>Name</Label>
+          <InputErrorsTooltip errors={nameErrors}>
+            <InputField
+              id={nameId}
+              value={name}
+              onChange={(event) => setName(event.target.value)}
+            />
+          </InputErrorsTooltip>
+        </Flex>
+        <Flex direction="column" css={{ gap: theme.spacing[3] }}>
+          <Label>Value</Label>
+          <InputErrorsTooltip errors={valueErrors}>
+            <div>
+              <CodeEditor
+                variables={variables}
+                defaultValue={value}
+                onChange={setValue}
+              />
+            </div>
+          </InputErrorsTooltip>
+        </Flex>
+        <Flex justify="end" css={{ gap: theme.spacing[5] }}>
+          <Button color="neutral" onClick={onBack}>
+            Cancel
+          </Button>
+          <Button
+            onClick={() => {
+              if (/[a-zA-z_]+[a-zA-z0-9_]*/.test(name) === false) {
+                setNameErrors([
+                  `Variable name can star with letter or _ and contain letters, digits and underscores`,
+                ]);
+                return;
+              }
+              const result = saveVariable(variable?.id, name, value);
+              if (result?.error !== undefined) {
+                setValueErrors([result.error]);
+                return;
+              }
+              onBack();
+            }}
+          >
+            Save
+          </Button>
+        </Flex>
+      </Flex>
+    </ScrollArea>
+  );
+};
+
+const selectedInstanceVariables = computed(
+  [selectedInstanceSelectorStore, dataSourcesStore],
+  (instanceSelector, dataSources) => {
+    const matchedVariables: VariableDataSource[] = [];
+    if (instanceSelector === undefined) {
+      return matchedVariables;
+    }
+    for (const dataSource of dataSources.values()) {
+      if (
+        dataSource.type === "variable" &&
+        dataSource.scopeInstanceId !== undefined &&
+        instanceSelector.includes(dataSource.scopeInstanceId)
+      ) {
+        matchedVariables.push(dataSource);
+      }
+    }
+    return matchedVariables;
+  }
+);
+
+const EmptyList = ({ onAdd }: { onAdd: () => void }) => {
+  return (
+    <Flex direction="column" css={{ gap: theme.spacing[5] }}>
+      <Flex justify="center" align="center" css={{ height: theme.spacing[13] }}>
+        <Text variant="labelsSentenceCase">No variables yet</Text>
+      </Flex>
+      <Flex justify="center" align="center" css={{ height: theme.spacing[13] }}>
+        <Button prefix={<PlusIcon />} onClick={onAdd}>
+          Create variable
+        </Button>
+      </Flex>
+    </Flex>
+  );
+};
+
+const deleteVariable = (variable: VariableDataSource) => {
+  // @todo prevent delete when variable is used
```

Comment:
probably also needs a confirmation dialog when trying to delete a variable that is used?

Reply:
I'd better disable button to let user clear everything self.